src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexTreeAnalyzer.cs

```diff
@@ -189,6 +203,26 @@ internal sealed class AnalysisResults
         /// </remarks>
         public bool MayBacktrack(RegexNode node) => !_complete || (_mayBacktrack?.Contains(node) ?? false);
 
+        /// <summary>Gets whether a node may be contained inside of one or more loops.</summary>
+        /// <remarks>
+        /// Constructs sometimes need to maintain state about their execution such that if they're backtracked
+        /// to, they have the necessary context to make a different choice and continue execution.  Such state
+        /// can often be maintained in locals dedicated to that construct.  If, however, the construct is inside
+        /// of a loop, an additional iteration of that outerloop could invoke the inner construct and cause those
+        /// locals to have their state overwritten.  In such situations, the constructs can't rely on the locals
+        /// maintaining the state and instead need to push the state on to a stack.  That pushing/popping has
+        /// additional cost, however, both in terms of run-time overheads and in terms of the additional code
+        /// required to handle it.  Code generators then can consult this <see cref="IsInLoop"/> to determine
+        /// whether locals are sufficient to maintain the state or whether the state needs to be pushed on to a stack.
+        ///
+        /// Loops are not considered to be "in" themselves.  This will only return true for a loop node if it's
+        /// nested inside of another loop node.
+        ///
+        /// If the whole tree couldn't be examined, this returns true.  That could lead to additional
+        /// backtracking-related code being emitted, but it's functionally the safe choice.
+        /// </remarks>
+        public bool IsInLoop(RegexNode node) => !_complete || (_inLoops?.Contains(node) ?? false);
```

Comment:
Just to make sure I understand here, the reason why we need this extra IsInLoop as opposed to using `MyBacktrack` to see if we can use locals, is because there are more cases where MayBacktrack returns true but IsInLoop might be false, for instance, a pattern like `(a|ab)c` may backtrack for strings like "abc" but should not require to keep around state as it is safe to abandon the first alternation state after not finding a `c` in the second position. Is that right?

Reply:
Not quite.  They're closer to opposites than they are to synonyms. MayBacktrack is about what a node or its children may do, whereas IsInLoop is about what the ancestors of a node may do.

Let's say you have the pattern that contains a repeater, e.g. `(...){10}`.  If the body of that loop can't backtrack, then this can simply be output as a loop to do the match 10 times. Easy. But if the body of the loop _can_ backtrack, now we need to handle the possibility that we, say, greedily match all 10 iterations, but then what comes after the loop fails, and backtracks into the last iteration, and it backtracks into the ninth iteration, and so on, and then if something matches again we need to re-greedily consume back up to 10.  By checking MayBacktrack, if it returns false we can emit the simple thing rather than the much more complicated logic.  Other examples involve atomic.  If an atomic or a lookaround or an expression conditional's expression wrap something that can't backtrack, they don't have to hide any contained backtracking goo from things that might come after, but if the they do, they need to deal with the fact that there's possibly backtracking state left on the stack, and need to be able to wipe it clean.

IsInLoop is more upwards looking than downwards looking.  Let's say we have `([ab]|[bc])b`, and let's say the alternation is storing its current branch in a local dedicated to this alternation.  No one else knows about or will touch that local.  We're matching against "aa".  The alternation will store 0 in that local after matching 'a' and will then proceed to the rest of the expression to try to match 'b'.  When it backtracks, we know that the local still contains 0; no one else could have overwritten it, because there's a 1:1 correspondence between this alternation construct and any locals it's using for state.  But now let's change the pattern to be `([ab]|[bc])*b`.  Now the alternation is in a loop.  That 1:1 correspondence no longer exists, because every iteration of the loop can have a different set of state that needs to be persisted in the locals, yet the locals can only store one value at a time.  So if we match against, say, "bca", we'll match the 'b' with the first iteration and the current branch will be set to 0, then we'll try to match another iteration, and we'll match the 'c' with the second branch, so current branch will be set to 1... if we only had the single local, we've now lost our branch information from the first iteration.  So, for constructs that use locals to persist state, if and only if they're in a loop, they need to persist their state to the stack so that values for each iteration can be stored and appropriately restored.

Reply:
I see, ok that makes more sense now, thanks a bunch for explaining